### PR TITLE
[tcgc] Method Input Signatures Doc

### PR DESCRIPTION
**Pull request by iscai-msft**

### CODE REVIEW

File: docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+</TabItem>
+</Tabs>
+
+### Anonymous Model
+
+<Tabs>
+<TabItem value="typespec" label="TypeSpec" default>
+
+```typespec
+op upload(firstName: string, lastName: string): void;
```
Comment (tadelesh) [inline docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx:180]: same with `op upload({firstName: string, lastName: string}): void;`
but not same with ` op upload(@path id: string, firstName: string, lastName: string): void;`

File: docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+<Tabs>
+<TabItem value="typespec" label="TypeSpec" default>
+
+```typespec
+alias User = {
+  firstName: string,
+  lastName: string,
+};
+
+op upload(...User): void;
```
Comment (tadelesh) [inline docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx:232]: for me, it is same as anonymous model spread. i wonder can we arrange the cases with more essential way. i mean:
1. `op update(param1, param2): void` raw parameter without decorator
2. `op update(@path path1, param1): void` raw parameters with decorator
3. `op update(...ModelWithoutDecoratedProperties): void` spread model without decorated properties
4. `op update(...ModelWithDecoratedProperties): void` spread model with decorated properties
5. `op update(...Alias/AnonymouModelWith/WithoutDecoratedProperties): void` spread anonymous model with/without decorated properties

Comment (iscai-msft) [reply docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx:232]: i'll switch the raw parameters to alias bc you are right, they are the same

**Comment by tadelesh** [comment]
how should we deal with model mixed with http decorator? for example:
```
model Test {
  @path
  prop1: string;
  @path
  prop2: string;
  prop3: string;
}
op test(...Test): void;
```
if we don't want to spread `Test` model, then we need to keep `prop1` and `prop2` in the `Test` model and do the mapping when handling request. then we could separate method and protocol clean.
if we want to put path parameter as method parameter, and generate model without path property, then we need to use protocol information when exporting method signature which make our initial design broken.

**Comment by weidongxu-microsoft** [comment]
Java would like to add a comment "use with caution". @srnagar 

Condition
1. the anonymous model should only contain very small number of properties (4 maybe?)
2. the anonymous model should likely stay same across versions (each optional property added in api-version would result in a new overload of the API in Java SDK)
3. the anonymous model cannot be used in json-merge-patch
4. the model (that now be spread to op) cannot be polymorphic (<-- this one probably be cross-language)

For 4, I mean this
```ts
@discriminator("kind")
model ModelBase {
  kind: string;
}

model Model extends ModelBase {
  kind: "abc";
  prop1: string;
}

@post
op post(...ModelBase, prop2: string): OkResponse;
```

File: docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+
+## Basic
+
+### HTTP Get
+
+<Tabs>
+<TabItem value="typespec" label="TypeSpec" default>
+
+```typespec
+op @get get(): User;
```
Comment (weidongxu-microsoft) [inline docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx:16]: All HTTP method decorator such as `@get` should be before `op`.

Comment (weidongxu-microsoft) [inline docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx:16]: All HTTP method decorator such as `@get` should be before `op`.

```suggestion
@get op get(): User;
```

File: docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+</TabItem>
+</Tabs>
+
+### HTTP Post
+
+<Tabs>
+<TabItem value="typespec" label="TypeSpec" default>
+
+```typespec
+op @post post(User): void;
```
Comment (weidongxu-microsoft) [inline docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx:61]:
```suggestion
@post op post(@body user: User): void;
```

File: docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+```python
+class User:
+  first_name: str
+  last_name: str
+
+class UserRequest:
+  user: User
+
+def upload(userRequest: UserRequest) -> None:
+  ...
```
Comment (weidongxu-microsoft) [inline docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx:308]: Not sure if Python code is correct.

If `@body user: User`, the request body is the `User` model.

File: docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+<TabItem value="typespec" label="TypeSpec" default>
+
+```typespec
+model BlobProperties {
+  @key
+  @visibility("read")
+  @path name: string;
+
+  @header contentLength: string;
+  @header contentType: string;
```
Comment (weidongxu-microsoft) [inline docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx:350]: tsp error

```
contentType parameter must be a string literal or union of string literalsTypeSpec(@typespec/http/content-type-string)
```

File: docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx (hunk 5 above)
Comment (msyyc) [reply docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx:308]: @iscai-msft The code seems not correct and here is the generated code with latest typespec-emitter:
image

File: docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx [hunk 7]
```diff
@@ -0,0 +1,456 @@
...
+}
+
+op setBlobProperties(...BlobProperties): void;
```
Comment (weidongxu-microsoft) [inline docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx:353]: route need to define the `name` as it is `@path`.

File: docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx (hunk 6 above)
Comment (weidongxu-microsoft) [reply docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx:350]: I would recommend not using multiple-content-type as example here.

File: docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx (hunk 4 above)
Comment (chunyu3) [reply docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx:65]: with/without @body decorator should have different behavior.  Does this part want to show the scenaro without @body decorator which mean anonymous body?

Comment (weidongxu-microsoft) [reply docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx:65]: I just mean `post(User)` is not valid in typespec... I just guess Isabella's intended code, not sure.

File: docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx [hunk 8]
```diff
@@ -0,0 +1,513 @@
...
+
+### Model with Decorated Properties
+
+<Tabs>
+<TabItem value="typespec" label="TypeSpec" default>
+
+```typespec
+model BlobProperties {
+  @key
+  @visibility("read")
```
Comment (chunyu3) [inline docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx:388]: Can `@visibility("read")` decorate a path parameter together with `@key` decorator? 
```suggestion
```

File: docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx [hunk 9]
```diff
@@ -0,0 +1,567 @@
...
+</TabItem>
+
+<TabItem value="python" label="Python">
+
+```python
+class User:
+  first_name: str
+  last_name: str
+
+def upload(user: User) -> None:
```
Comment (msyyc) [inline docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx:280]: if no explicit `@body user: User`, the generated python code will be
```
    def upload(  # pylint: disable=inconsistent-return-statements
        self, body: Union[_models.User, JSON, IO[bytes]], **kwargs: Any
    ) -> None:
```

File: docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx (hunk 8 above)
Comment (msyyc) [reply docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx:430]: It is permitted: 
image

Comment (msyyc) [reply docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx:430]: nit: it seems `@path` is not necessary if it is already decorated with `@key`

File: docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx (hunk 4 above)
Comment (tadelesh) [reply docs/howtos/DataPlane Generation - DPG/08methodInputs.mdx:70]: this case aims to show the basic usage, so weidong is right, i changed the doc.

**Comment by qiaozha** [comment]
js pending #712

**Review by lmazuel — APPROVED** [review]
All languages confirmed they have a design, by pushing code to it. Approving.

**Pull request merged**